io/net/net_tools.py

```diff
@@ -0,0 +1,289 @@
+                self.error("traceroute6 did not show 1 hop for localhost6")
+
+    def test_route(self):
+        """
+        Verify the route fuctionality.
+        """
+        output = process.run(cmd="route -n",
+                                 shell=True,
+                                 verbose=False)
+        if output.exit_status != 0:
+            self.error("route cmd failed")
+        if self.ipv6:
+            output = process.run(cmd="route -A inet6 -n",
+                                     shell=True,
+                                     verbose=False)
+            if output.exit_status != 0:
+                self.error("route ipv6 failed")
+
+    def test_ipmaddr(self):
+        """
+        Verify ipmaddr functionality
+        """
+        output = process.run(cmd="ipmaddr show dev lo",
+                                 shell=True,
+                                 verbose=False)
+        if output.exit_status != 0:
+            self.error("ipmaddr failed to diaplay mcast addr of loop back")
+        if self.ipv6:
+            output = process.run(cmd="ipmaddr show ipv6 dev lo",
+                                     shell=True,
+                                     verbose=False)
+            if output.exit_status != 0:
+                self.error("\n ipmaddr ipv6 failed")
+
+    def test_iptunnel(self):
+        """
+        Verify iptunnel functionality
+        """
+        # Note, this messes up dhcp client configurations so is skipped in that
+        # case
+        output = process.system_output("ps -aef", verbose=False, shell=True)
+        if 'dhclient' in output:
+            self.log.info("Skipped on dhclient systems")
+            return 0
+        # add sit1
+        status = process.system(
+            cmd="iptunnel add sit1 mode sit local 127.0.0.1 ttl 64",
+            verbose=False,
+            shell=True)
+        if status != 0:
+            self.error("iptunnel add sit1 failed")
+        output = process.run(cmd="iptunnel show",
+                                 shell=True,
+                                 verbose=False)
+        if output.exit_status != 0:
+            self.error("iptunnel show failed")
+        if not re.search("sit1", output.stdout):
+            self.error("iptunnel didn't add sit1")
+        # remove sit1
+        status = process.system("iptunnel del sit1", verbose=False,
+                                shell=True)
+        if status != 0:
+            self.error("iptunnel del sit1 failed")
+        output = process.run(cmd="iptunnel show",
+                                 shell=True,
+                                 verbose=False)
+        if output.exit_status != 0:
+            self.error("iptunnel show failed")
+        if re.search("sit1", output.stdout):
+            self.error("iptunnel didn't remove sit1")
+
+    def teardown(self):
+        """
+        cleanup function
+        """
+        if os.path.exists(self.backup_dir):
```

Comment:
Well this is no cleanup. You change several setting in the tests and in case of failure you should return the system into the proper state here. It'd be really hard to do using this structure so I'd suggest splitting this test into multiple tests, each with custom `setUp` and `tearDown` methods.

When you're changing any setting you first in `setUp` need to check if it's available and if the value is not the same you're setting it to. Then in test you should store the original value just before you change it (alternatively you can do it in `setUp`) and then change it. The `tearDown` method should check whether the value was changed and if so it should try to recover it. In case of failure it should report `tearDown` failure.

An example:

```python
def install_dependencies():
    distro.detect()
    if self.detected_distro.name == "SuSE":
        net_tools = ("net-tools", "traceroute")
    else:
        net_tools = ("net-tools", "hostname", "traceroute")
    for pkg in net_tools:
        if not sm.check_installed(pkg) and not sm.install(pkg):
            raise AssertionError("%s package is need to test" % pkg)


class HostName(test):
    ...


class IPTunnel(test):
    def setUp(self):
        self.tunnel = None
        if 'dhclient' in process.system_output("ps -aef", verbose=False):
            self.skip("Test not supported on systems running dhclient")
        install_dependencies()
        pre = process.system_output("iptunnel show")
        if "sit1" in pre:
            self.skip("'sit1' already configured in iptunnel: %s" % pre)
    def test_loopback_sit(self):
        self.tunnel = "sit1"    # We'll try to add it in the next step
        process.system("iptunnel add sit1 mode sit local 127.0.0.1 ttl 64",
                       root=True)
        ret = process.run("iptunnel show")
        if "sit1" not in ret.stdout:
            self.fail("sit1 not listed in:\n%s" % ret)
        ret = self._remove_tunnel("sit1")
        if ret:
            self.fail(ret)
        self.tunnel = None  # It's already removed, no need to do it again
    @staticmethod
    def _remove_tunnel(name):
        process.system("iptunnel del %s" % name, sudo=True)
        ret = process.run("iptunnel show")
        if name in ret.stdout:
            return "%s still in the list:\n%s" % (name, ret))
    def tearDown(self):
        if self.tunnel:
            ret = self._remove_tunnel(self.tunnel):
            if ret:
                self.error("Unable to clear tunnel %s: %s" % ret)

```